### PR TITLE
Downloader improvements and make offline development possible

### PR DESCRIPTION
**Pull request by char3210**

### CODE REVIEW

File: src/main/java/org/parchmentmc/compass/util/download/ManifestsDownloader.java [hunk 1]
```diff
@@ -72,7 +72,9 @@ private LauncherManifest downloadLauncherManifest() {
             Directory outputDir = this.outputDirectory.get();
             File outputFile = outputDir.file("launcher.json").getAsFile();
 
-            createAndExecuteAction(project, manifestURL, outputFile, "launcher manifest");
+            if (!outputFile.exists()) {
+                createAndExecuteAction(project, manifestURL, outputFile, "launcher manifest");
+            }
```
Comment (marchermans) [inline src/main/java/org/parchmentmc/compass/util/download/ManifestsDownloader.java]: This is not a good idea, at least with regards to the launcher manifest, it should be downloaded at all times, only if gradle is offline should it be skipped.

**Review by marchermans — CHANGES_REQUESTED** [review]

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]
In addition to @marchermans's point, our download utility is backed by the [gradle-download-task](https://github.com/michel-kraemer/gradle-download-task) plugin, which I believe supports Gradle's native offline mode (`--offline` through the CLI; Eclipse and IntelliJ support it too).

The reason I didn't put those output file existence checks when writing the download utility originally is because that would mean needing to manually delete the manifests when they change remotely, as otherwise it would always skip the download action if the manifest exists. As you can see in the beginning of [`DownloadUtil#createAndExecuteAction`](https://github.com/ParchmentMC/Compass/blob/d1416a404d89ad3992776e121715fcc7aad7b4b7/src/main/java/org/parchmentmc/compass/util/download/DownloadUtil.java#L19-L21), it is configured to always check the remote source for changes (using ETags) and download only if there are changes.

The only change I have no hesitance about is the change in `DownloadUtil#createAndExecuteAction` that makes it catch all exceptions and appends the output path to the exception message.

**Comment by char3210** [comment]
yeah that makes sense, for some reason I was having some trouble having it work with intellij's offline mode. should the check for `VersionDownload` still be checked though, since it checks the hash?

**Review by sciwhiz12 — APPROVED** [review]

**Comment by sciwhiz12** [comment]
> should the check for `VersionDownload` still be checked though, since it checks the hash?

Sorry for not addressing this earlier. The check should be fine, as the hash of the file in the version manifest should only really change if the file changes on the server, which we then redownload. When the hash doesn't change (which is 99% the scenario), the file didn't change, so no need to query the server for that same info.

**Pull request merged**